src/env.h

```diff
@@ -1172,13 +1198,14 @@ class Environment {
   void RunAndClearNativeImmediates();
   static void CheckImmediate(uv_check_t* handle);
 
-  struct CleanupHookCallback {
-    void (*fn_)(void*);
-    void* arg_;
-
-    // We keep track of the insertion order for these objects, so that we can
-    // call the callbacks in reverse order when we are cleaning up.
-    uint64_t insertion_order_counter_;
+  class CleanupHookCallback : public MemoryRetainer {
```

Comment:
I’m a bit worried about the overhead that this comes with… it means 1 extra pointer field for each cleanup hook instance.

If the goal is to extend this to more classes, how moving `DeleteMe` from `BaseObject` to `MemoryRetainer`, and allow other subclasses of `MemoryRetainer` to use that as well?

Alternatively, it might make sense to put `BaseObject`s/`MemoryRetainer`s into a separate data structure that doesn’t need the `fn_` field at all, or use an intrusive linked list (like `HandleWrapQueue`)?

Reply:
I believe an easier (and cleaner?) alternative would be, instead of having the class itself inherit from `MemoryRetainer`, create a specialized `TrackField` (like what we do for v8 handles and libuv handles) that track those fields properly, which might be preferable in this case because  no one actually inherits from `CleanupHookCallback` (we can also mark it final to prevent that).

Reply:
Yeah, I think that should work as well :+1: Maybe use SFINAE to detect whether the class we’re tracking has a `TrackField()` method, like we do for the (currently unused?) overload of `MemoryRetainer::TrackField` that has a `test_for_number` template parameter?

Reply:
> If the goal is to extend this to more classes, how moving DeleteMe from BaseObject to MemoryRetainer, and allow other subclasses of MemoryRetainer to use that as well?

Off the top of my head, this seems feasible, but that would make `MemoryRetainer` more sophisticated than what it currently is - it is now only for the heap snapshot integration, moving memory management bits there may require a bit more thoughts into this...(the general idea of making it more powerful sounds cool though, and probably can help with https://github.com/nodejs/diagnostics/issues/155)

> Alternatively, it might make sense to put BaseObjects/MemoryRetainers into a separate data structure that doesn’t need the fn_ field at all, or use an intrusive linked list (like HandleWrapQueue)?

That's also what I wondered when I implemented this (although, I decided to defer implementing support for our internal linked list in a future patch, because I do not want to make this patch too big).

Reply:
> Yeah, I think that should work as well 👍 Maybe use SFINAE to detect whether the class we’re tracking has a TrackField() method, like we do for the (currently unused?) overload of MemoryRetainer::TrackField that has a test_for_number template parameter?

We could just create an overload for `CleanupHookCallback` instead of doing any template magic. (BTW the test_for_number variant is used and is why this patch does not track the STL maps that have numeric keys - then you'll see funny pairs with a value as "second" but without "first" because the key is elided into the pair).